File: FretTrainerEZ/Game/GameState.swift
import Foundation
import CoreHaptics
import Observation

enum GameMode: String, CaseIterable {
    case nameTheNote     = "Name That Note"
    case findTheFret     = "Find The Fret"
    case memoryChallenge = "Memory Challenge"

    var shortName: String {
        switch self {
        case .nameTheNote:     return "Name It"
        case .findTheFret:     return "Find It"
        case .memoryChallenge: return "Memory"
        }
    }
}

enum MemoryPhase: Equatable {
    case flashing    // all positions shown on board
    case recalling   // board cleared — user taps from memory
    case complete    // brief pause revealing missed spots before next round
}

struct FretPosition: Hashable {
    let string: Int
    let fret: Int
}

enum Difficulty: String, CaseIterable {
    case beginner     = "Beginner"
    case intermediate = "Intermediate"
    case advanced     = "Advanced"

    var maxFret: Int {
        switch self {
        case .beginner:     return 5
        case .intermediate: return 10
        case .advanced:     return 22
        }
    }
}

@Observable
final class GameState {
    private let fretboard: Fretboard

    // Difficulty
    var difficulty: Difficulty = .beginner

    var gameMode: GameMode = .nameTheNote

    // Current question
    var currentString: Int = 0
    var currentFret: Int = 0
    var correctNote: Note = .C
    var questionID: UUID = UUID()

    // Score
    var correctCount: Int = 0
    var totalCount: Int = 0

    var scorePercent: Int {
        guard totalCount > 0 else { return 0 }
        return Int(Double(correctCount) / Double(totalCount) * 100)
    }

    // Streak
    var currentStreak: Int = 0
    var bestStreakThisSession: Int = 0
    var bestStreak: Int {
        get { UserDefaults.standard.integer(forKey: bestStreakKey) }
    }
    private var bestStreakKey: String { "bestStreak_\(gameMode.rawValue)" }

    // Timed session result
    var showTimedResult: Bool = false

    // Answer feedback
    enum AnswerState: Equatable {
        case idle
        case correct(tapped: Note)
        case wrong(tapped: Note, correct: Note)
    }
    var answerState: AnswerState = .idle

    enum FretAnswerState: Equatable {
        case idle
        case correct(string: Int, fret: Int)
        case wrong(string: Int, fret: Int)
    }
    var fretAnswerState: FretAnswerState = .idle

    // Find The Fret / Memory: positions the user has already tapped correctly this round
    var foundFrets: Set<FretPosition> = []

    // Memory Challenge
    var memoryPhase: MemoryPhase = .flashing
    private var memoryFlashTimer: Timer?

    var flashDuration: Double {
        switch difficulty {
        case .beginner:     return 4.0
        case .intermediate: return 2.5
        case .advanced:     return 1.5
        }
    }

    /// All valid positions for the current note within the active difficulty range.
    var required: Set<FretPosition> {
        Set(fretboard.allPositions(for: correctNote)
            .filter { $0.fret <= difficulty.maxFret }
            .map { FretPosition(string: $0.string, fret: $0.fret) })
    }

    // Best score persistence
    var isNewBest: Bool = false
    var bestTimedScore: Int { UserDefaults.standard.integer(forKey: timedScoreKey) }
    private var timedScoreKey: String { "best_\(gameMode.rawValue)_\(timerDuration)" }

    // Timer
    var isTimedMode: Bool = false
    var timerDuration: Int = 60
    var timeRemaining: Int = 60
    var isTimerActive: Bool = false
    var isTimeUp: Bool = false

    var canAnswer: Bool {
        if isTimedMode { return isTimerActive && !isTimeUp }
        return true
    }

    private var countdownTimer: Timer?
    private var hapticEngine: CHHapticEngine?

    init(fretboard: Fretboard = Fretboard()) {
        self.fretboard = fretboard
        prepareHaptics()
        nextQuestion()
    }

    func nextQuestion() {
        answerState    = .idle
        fretAnswerState = .idle
        foundFrets     = []
        questionID     = UUID()
        switch gameMode {
        case .nameTheNote:
            currentString = Int.random(in: 0..<fretboard.tuning.stringCount)
            currentFret   = Int.random(in: 0...difficulty.maxFret)
            correctNote   = fretboard.note(string: currentString, fret: currentFret)
        case .findTheFret:
            correctNote   = Note.allCases.randomElement()!
            currentString = 0
            currentFret   = 0
        case .memoryChallenge:
            let otherNotes = Note.allCases.filter { $0 != correctNote }
            correctNote   = otherNotes.randomElement() ?? correctNote
            currentString = 0
            currentFret   = 0
            scheduleMemoryFlash()
        }
    }

    private func scheduleMemoryFlash() {
        memoryPhase = .flashing
        memoryFlashTimer?.invalidate()
        memoryFlashTimer = Timer.scheduledTimer(withTimeInterval: flashDuration, repeats: false) { [weak self] _ in
            guard let self else { return }
            self.memoryPhase    = .recalling
            self.memoryFlashTimer = nil
        }
    }

    func setDifficulty(_ newDifficulty: Difficulty) {
        difficulty = newDifficulty
        correctCount = 0
        totalCount = 0
        nextQuestion()
    }

    func submit(answer: Note) {
        guard case .idle = answerState else { return }
        guard canAnswer else { return }
        totalCount += 1
        let correctDelay: Double = isTimedMode ? 0.5 : 0.8
        let wrongDelay: Double   = isTimedMode ? 0.7 : 1.5
        if answer == correctNote {
            correctCount += 1
            currentStreak += 1
            if currentStreak > bestStreakThisSession { bestStreakThisSession = currentStreak }
            answerState = .correct(tapped: answer)
            playHaptic(success: true)
            DispatchQueue.main.asyncAfter(deadline: .now() + correctDelay) { [weak self] in
                guard let self, !self.isTimeUp else { return }
                self.nextQuestion()
            }
        } else {
            currentStreak = 0
            answerState = .wrong(tapped: answer, correct: correctNote)
            playHaptic(success: false)
            DispatchQueue.main.asyncAfter(deadline: .now() + wrongDelay) { [weak self] in
                guard let self, !self.isTimeUp else { return }
                self.nextQuestion()
            }
        }
    }

    func submitFret(string: Int, fret: Int) {
        guard canAnswer else { return }
        let pos = FretPosition(string: string, fret: fret)
        guard !foundFrets.contains(pos) else { return }   // already lit up
        guard case .idle = fretAnswerState else { return } // block during wrong flash / completion

        if fretboard.note(string: string, fret: fret) == correctNote {
            foundFrets.insert(pos)
            playHaptic(success: true)

            // Advance when all required positions (within difficulty range) are found
            if required.isSubset(of: foundFrets) {
                correctCount += 1
                totalCount += 1
                currentStreak += 1
                if currentStreak > bestStreakThisSession { bestStreakThisSession = currentStreak }
                fretAnswerState = .correct(string: string, fret: fret) // blocks further taps
                DispatchQueue.main.asyncAfter(deadline: .now() + (isTimedMode ? 0.5 : 0.8)) { [weak self] in
                    guard let self, !self.isTimeUp else { return }
                    self.nextQuestion()
                }
            }
        } else {
            totalCount += 1
            currentStreak = 0
            fretAnswerState = .wrong(string: string, fret: fret)
            playHaptic(success: false)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) { [weak self] in
                guard let self else { return }
                if case .wrong = self.fretAnswerState { self.fretAnswerState = .idle }
            }
        }
    }

    func submitMemoryTap(string: Int, fret: Int) {
        guard memoryPhase == .recalling else { return }
        let pos = FretPosition(string: string, fret: fret)
        guard !foundFrets.contains(pos) else { return }
        guard case .idle = fretAnswerState else { return }

        if fretboard.note(string: string, fret: fret) == correctNote {
            foundFrets.insert(pos)
            totalCount   += 1
            correctCount += 1
            currentStreak += 1
            if currentStreak > bestStreakThisSession { bestStreakThisSession = currentStreak }
            playHaptic(success: true)

            if required.isSubset(of: foundFrets) {
                memoryPhase = .complete
                DispatchQueue.main.asyncAfter(deadline: .now() + 1.2) { [weak self] in
                    guard let self else { return }
                    self.nextQuestion()
                }
            }
        } else {
            totalCount    += 1
            currentStreak  = 0
            fretAnswerState = .wrong(string: string, fret: fret)
            playHaptic(success: false)
            DispatchQueue.main.asyncAfter(deadline: .now() + 0.6) { [weak self] in
                guard let self else { return }
                if case .wrong = self.fretAnswerState { self.fretAnswerState = .idle }
            }
        }
    }

    func skipNote() {
        guard gameMode == .findTheFret else { return }
        let other = Note.allCases.filter { $0 != correctNote }
        correctNote = other.randomElement() ?? correctNote
        answerState = .idle
        fretAnswerState = .idle
        foundFrets = []
    }

    func setGameMode(_ mode: GameMode) {
        gameMode = mode
        reset()
    }

    func reset() {
        memoryFlashTimer?.invalidate()
        memoryFlashTimer = nil
        memoryPhase     = .flashing
        fretAnswerState = .idle
        foundFrets      = []
        isNewBest       = false
        showTimedResult = false
        currentStreak   = 0
        bestStreakThisSession = 0
        stopTimedGame()
        correctCount = 0
        totalCount   = 0
        nextQuestion()
    }

    func startTimedGame() {
        correctCount = 0
        totalCount = 0
        isNewBest = false
        showTimedResult = false
        currentStreak = 0
        bestStreakThisSession = 0
        foundFrets = []
        timeRemaining = timerDuration
        isTimeUp = false
        isTimerActive = true
        nextQuestion()
        countdownTimer?.invalidate()
        countdownTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { [weak self] _ in
            guard let self else { return }
            if self.timeRemaining > 1 {
                self.timeRemaining -= 1
            } else {
                self.timeRemaining = 0
                self.isTimerActive = false
                self.isTimeUp = true
                self.answerState = .idle
                self.countdownTimer?.invalidate()
                self.countdownTimer = nil
                self.saveTimedScoreIfBetter()
            }
        }
    }

    private func saveTimedScoreIfBetter() {
        if correctCount > bestTimedScore {
            UserDefaults.standard.set(correctCount, forKey: timedScoreKey)
            isNewBest = true
        }
        if bestStreakThisSession > bestStreak {
            UserDefaults.standard.set(bestStreakThisSession, forKey: bestStreakKey)
        }
        showTimedResult = true
    }

    func stopTimedGame() {
        countdownTimer?.invalidate()
        countdownTimer = nil
        isTimerActive = false
        isTimeUp = false
        timeRemaining = timerDuration
    }

    // MARK: - Haptics
    private func prepareHaptics() {
        guard CHHapticEngine.capabilitiesForHardware().supportsHaptics else { return }
        hapticEngine = try? CHHapticEngine()
        try? hapticEngine?.start()
    }

    private func playHaptic(success: Bool) {
        guard UserDefaults.standard.object(forKey: "hapticsEnabled") as? Bool ?? true else { return }
        guard let engine = hapticEngine,
              CHHapticEngine.capabilitiesForHardware().supportsHaptics else { return }
        let intensity = CHHapticEventParameter(parameterID: .hapticIntensity, value: success ? 0.8 : 1.0)
        let sharpness = CHHapticEventParameter(parameterID: .hapticSharpness, value: success ? 0.3 : 0.9)
        let event = CHHapticEvent(eventType: success ? .hapticContinuous : .hapticTransient,
                                  parameters: [intensity, sharpness],
                                  relativeTime: 0,
                                  duration: success ? 0.3 : 0.1)
        if let pattern = try? CHHapticPattern(events: [event], parameters: []),
           let player = try? engine.makePlayer(with: pattern) {
            try? player.start(atTime: CHHapticTimeImmediate)
        }
    }
}

File: android/app/src/main/kotlin/com/dontfretaboutitai/frettrainerez/game/GameState.kt
package com.dontfretaboutitai.frettrainerez.game

import android.app.Application
import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.dontfretaboutitai.frettrainerez.models.Fretboard
import com.dontfretaboutitai.frettrainerez.models.FretPosition
import com.dontfretaboutitai.frettrainerez.models.Note
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

// ---------------------------------------------------------------------------
// Supporting enums
// ---------------------------------------------------------------------------

enum class GameMode(val displayName: String, val shortName: String) {
    NAME_THE_NOTE("Name That Note", "Name It"),
    FIND_THE_FRET("Find The Fret", "Find It"),
    MEMORY_CHALLENGE("Memory Challenge", "Memory"),
}

enum class Difficulty(val displayName: String, val maxFret: Int) {
    BEGINNER("Beginner", 5),
    INTERMEDIATE("Intermediate", 10),
    ADVANCED("Advanced", 22),
}

enum class MemoryPhase { FLASHING, RECALLING, COMPLETE }

sealed class AnswerState {
    object Idle : AnswerState()
    data class Correct(val tapped: Note) : AnswerState()
    data class Wrong(val tapped: Note, val correct: Note) : AnswerState()
}

sealed class FretAnswerState {
    object Idle : FretAnswerState()
    data class Correct(val string: Int, val fret: Int) : FretAnswerState()
    data class Wrong(val string: Int, val fret: Int) : FretAnswerState()
}

// ---------------------------------------------------------------------------
// GameState ViewModel
// ---------------------------------------------------------------------------

class GameState(application: Application) : AndroidViewModel(application) {

    val fretboard = Fretboard()
    private val prefs = application.getSharedPreferences("fret_trainer", Context.MODE_PRIVATE)

    // -- Mode & difficulty --
    var gameMode by mutableStateOf(GameMode.NAME_THE_NOTE)
    var difficulty by mutableStateOf(Difficulty.BEGINNER)

    // -- Current question --
    var currentString by mutableStateOf(0)
    var currentFret   by mutableStateOf(0)
    var correctNote   by mutableStateOf(Note.C)
    var questionIndex by mutableStateOf(0)   // incremented on each new question (drives sound trigger)

    // -- Score --
    var correctCount by mutableStateOf(0)
    var totalCount   by mutableStateOf(0)
    val scorePercent: Int get() = if (totalCount > 0) correctCount * 100 / totalCount else 0

    // -- Streak --
    var currentStreak        by mutableStateOf(0)
    var bestStreakThisSession by mutableStateOf(0)

    val bestStreak: Int get() = prefs.getInt(bestStreakKey, 0)
    private val bestStreakKey: String get() = "bestStreak_${gameMode.name}"

    // -- Answer states --
    var answerState     by mutableStateOf<AnswerState>(AnswerState.Idle)
    var fretAnswerState by mutableStateOf<FretAnswerState>(FretAnswerState.Idle)

    // -- Multi-tap tracking (Find The Fret + Memory) --
    var foundFrets by mutableStateOf(emptySet<FretPosition>())

    // -- Memory Challenge --
    var memoryPhase by mutableStateOf(MemoryPhase.FLASHING)

    val flashDuration: Long get() = when (difficulty) {
        Difficulty.BEGINNER     -> 4000L
        Difficulty.INTERMEDIATE -> 2500L
        Difficulty.ADVANCED     -> 1500L
    }

    /** All valid positions for the current note within the active difficulty range. */
    val required: Set<FretPosition> get() =
        fretboard.allPositionsFor(correctNote)
            .filter { it.fret <= difficulty.maxFret }
            .toSet()

    // -- Timed mode --
    var isTimedMode     by mutableStateOf(false)
    var timerDuration   by mutableStateOf(60)
    var timeRemaining   by mutableStateOf(60)
    var isTimerActive   by mutableStateOf(false)
    var isTimeUp        by mutableStateOf(false)
    var isNewBest       by mutableStateOf(false)
    var showTimedResult by mutableStateOf(false)

    val canAnswer: Boolean get() = if (isTimedMode) (isTimerActive && !isTimeUp) else true

    val bestTimedScore: Int get() = prefs.getInt(timedScoreKey, 0)
    private val timedScoreKey: String get() = "best_${gameMode.name}_$timerDuration"

    /** Returns the best timed score for the given duration in the current game mode. */
    fun bestScoreFor(duration: Int): Int = prefs.getInt("best_${gameMode.name}_$duration", 0)

    private var countdownJob:   Job? = null
    private var memoryFlashJob: Job? = null

    // ---------------------------------------------------------------------------
    // Init
    // ---------------------------------------------------------------------------

    init {
        nextQuestion()
    }

    // ---------------------------------------------------------------------------
    // Mode & Difficulty
    // ---------------------------------------------------------------------------

    fun changeMode(mode: GameMode) {
        gameMode = mode
        reset()
    }

    fun changeDifficulty(d: Difficulty) {
        difficulty   = d
        correctCount = 0
        totalCount   = 0
        nextQuestion()
    }

    // ---------------------------------------------------------------------------
    // Question Generation
    // ---------------------------------------------------------------------------

    fun nextQuestion() {
        answerState     = AnswerState.Idle
        fretAnswerState = FretAnswerState.Idle
        foundFrets      = emptySet()
        questionIndex  += 1

        when (gameMode) {
            GameMode.NAME_THE_NOTE -> {
                currentString = (0 until fretboard.tuning.stringCount).random()
                currentFret   = (0..difficulty.maxFret).random()
                correctNote   = fretboard.note(currentString, currentFret)
            }
            GameMode.FIND_THE_FRET -> {
                correctNote   = Note.entries.random()
                currentString = 0
                currentFret   = 0
            }
            GameMode.MEMORY_CHALLENGE -> {
                val others  = Note.entries.filter { it != correctNote }
                correctNote = others.random()
                currentString = 0
                currentFret   = 0
                scheduleMemoryFlash()
            }
        }
    }

    private fun scheduleMemoryFlash() {
        memoryPhase = MemoryPhase.FLASHING
        memoryFlashJob?.cancel()
        memoryFlashJob = viewModelScope.launch {
            delay(flashDuration)
            memoryPhase    = MemoryPhase.RECALLING
            memoryFlashJob = null
        }
    }

    // ---------------------------------------------------------------------------
    // Name That Note
    // ---------------------------------------------------------------------------

    fun submit(answer: Note) {
        if (answerState != AnswerState.Idle || !canAnswer) return
        totalCount += 1

        if (answer == correctNote) {
            correctCount  += 1
            currentStreak += 1
            if (currentStreak > bestStreakThisSession) bestStreakThisSession = currentStreak
            answerState = AnswerState.Correct(tapped = answer)
            val delay = if (isTimedMode) 500L else 800L
            viewModelScope.launch {
                delay(delay)
                if (!isTimeUp) nextQuestion()
            }
        } else {
            currentStreak = 0
            answerState   = AnswerState.Wrong(tapped = answer, correct = correctNote)
            val delay = if (isTimedMode) 700L else 1500L
            viewModelScope.launch {
                delay(delay)
                if (!isTimeUp) nextQuestion()
            }
        }
    }

    // ---------------------------------------------------------------------------
    // Find The Fret
    // ---------------------------------------------------------------------------

    fun submitFret(string: Int, fret: Int) {
        if (!canAnswer || fret > difficulty.maxFret) return
        val pos = FretPosition(string, fret)
        if (foundFrets.contains(pos) || fretAnswerState !is FretAnswerState.Idle) return

        if (fretboard.note(string, fret) == correctNote) {
            foundFrets = foundFrets + pos
            if (required.isSubset(foundFrets)) {
                correctCount  += 1
                totalCount    += 1
                currentStreak += 1
                if (currentStreak > bestStreakThisSession) bestStreakThisSession = currentStreak
                fretAnswerState = FretAnswerState.Correct(string, fret)
                val delay = if (isTimedMode) 500L else 800L
                viewModelScope.launch {
                    delay(delay)
                    if (!isTimeUp) nextQuestion()
                }
            }
        } else {
            totalCount      += 1
            currentStreak    = 0
            fretAnswerState  = FretAnswerState.Wrong(string, fret)
            viewModelScope.launch {
                delay(600L)
                if (fretAnswerState is FretAnswerState.Wrong) fretAnswerState = FretAnswerState.Idle
            }
        }
    }

    fun skipNote() {
        if (gameMode != GameMode.FIND_THE_FRET) return
        val others  = Note.entries.filter { it != correctNote }
        correctNote     = others.random()
        answerState     = AnswerState.Idle
        fretAnswerState = FretAnswerState.Idle
        foundFrets      = emptySet()
        questionIndex  += 1
    }

    // ---------------------------------------------------------------------------
    // Memory Challenge
    // ---------------------------------------------------------------------------

    fun submitMemoryTap(string: Int, fret: Int) {
        if (memoryPhase != MemoryPhase.RECALLING || fret > difficulty.maxFret) return
        val pos = FretPosition(string, fret)
        if (foundFrets.contains(pos) || fretAnswerState !is FretAnswerState.Idle) return

        if (fretboard.note(string, fret) == correctNote) {
            foundFrets = foundFrets + pos
            if (required.isSubset(foundFrets)) {
                correctCount  += 1
                totalCount    += 1
                currentStreak += 1
                if (currentStreak > bestStreakThisSession) bestStreakThisSession = currentStreak
                memoryPhase = MemoryPhase.COMPLETE
                viewModelScope.launch {
                    delay(1200L)
                    nextQuestion()
                }
            }
        } else {
            totalCount      += 1
            currentStreak    = 0
            fretAnswerState  = FretAnswerState.Wrong(string, fret)
            viewModelScope.launch {
                delay(600L)
                if (fretAnswerState is FretAnswerState.Wrong) fretAnswerState = FretAnswerState.Idle
            }
        }
    }

    // ---------------------------------------------------------------------------
    // Reset
    // ---------------------------------------------------------------------------

    fun reset() {
        memoryFlashJob?.cancel()
        memoryFlashJob        = null
        memoryPhase           = MemoryPhase.FLASHING
        answerState           = AnswerState.Idle
        fretAnswerState       = FretAnswerState.Idle
        foundFrets            = emptySet()
        correctCount          = 0
        totalCount            = 0
        currentStreak         = 0
        bestStreakThisSession  = 0
        isNewBest             = false
        showTimedResult       = false
        stopTimedGame()
        nextQuestion()
    }

    // ---------------------------------------------------------------------------
    // Timed Mode
    // ---------------------------------------------------------------------------

    fun startTimedGame() {
        correctCount         = 0
        totalCount           = 0
        currentStreak        = 0
        bestStreakThisSession = 0
        foundFrets           = emptySet()
        isNewBest            = false
        showTimedResult      = false
        timeRemaining        = timerDuration
        isTimeUp             = false
        isTimerActive        = true
        nextQuestion()
        countdownJob?.cancel()
        countdownJob = viewModelScope.launch {
            while (timeRemaining > 1) {
                delay(1000L)
                timeRemaining -= 1
            }
            delay(1000L)
            timeRemaining = 0
            isTimerActive = false
            isTimeUp      = true
            answerState   = AnswerState.Idle
            countdownJob  = null
            saveTimedScore()
        }
    }

    fun stopTimedGame() {
        countdownJob?.cancel()
        countdownJob  = null
        isTimerActive = false
        isTimeUp      = false
        timeRemaining = timerDuration
    }

    private fun saveTimedScore() {
        if (correctCount > bestTimedScore) {
            prefs.edit().putInt(timedScoreKey, correctCount).apply()
            isNewBest = true
        }
        if (bestStreakThisSession > bestStreak) {
            prefs.edit().putInt(bestStreakKey, bestStreakThisSession).apply()
        }
        showTimedResult = true
    }

    // ---------------------------------------------------------------------------
    // Helpers
    // ---------------------------------------------------------------------------

    private fun Set<FretPosition>.isSubset(other: Set<FretPosition>) = other.containsAll(this)
}
